### ISSUE TITLE
Label Blinking Issue at Integer Zoom Levels with animateCamera

### ISSUE THREAD

**Issue by geolives-contact** [maplibre/maplibre-native#2477]
**Describe the bug**
Starting with MapLibre v11, when the map zoom is set to an exact integer value (such as 15.0, 16.0 or 17.0 for instance), some labels blink when `animateCamera` is called.
The issue does not occur when using `moveCamera` or when the zoom is set between integer values (e.g., 16.4 or 17.1).

Here are some videos to demonstrate the issue:
* [First video]( https://drive.google.com/file/d/17WVr15KOfLPlXFBWX2onL0JVjPZUGXZR/view?usp=sharing): Shows the issue with labels for roads blinking during the animation.
* [Second video](https://drive.google.com/file/d/1UcCK5TUXKJh47Ra12Y4xhmaPRIhQAH9K/view?usp=sharing): Shows the labels not blinking when the zoom is set between integer values.

**To Reproduce**

1. Download and use this sample activity file: [SimpleMapActivity.kt.zip](https://github.com/user-attachments/files/15613146/SimpleMapActivity.kt.zip)
2. This activity file contains code to run animateCamera at short intervals with the zoom set exactly to 15.0.
3. If needed, we can create a branch with this edit in our fork.

**Expected behavior**
On the MapLibre v10 version, labels didn't blink when zoom are set at integer values.

**Platform information (please complete the following information):**
 - Platform: Android, iOS not tested
 - Version 11

**Comment by louwers** [comment]
Thanks for the great bug report with reproduction!

That's a really strange bug, but we'll have a look at it.

**Comment by louwers** [comment]
I tried iOS, doesn't seem to be a problem there:


```swift
struct SimpleMap: UIViewRepresentable {

    func makeUIView(context _: Context) -> MLNMapView {
        let mapView = MLNMapView(frame: .zero, styleURL: VERSATILES_COLORFUL_STYLE)
        mapView.zoomLevel = 15.0
        
        mapView.centerCoordinate = CLLocationCoordinate2D(latitude: 50.15, longitude: 5.2)
        
        startTimer(mapView: mapView)
        return mapView
    }

    func updateUIView(_: MLNMapView, context _: Context) {}
    
    private func startTimer(mapView: MLNMapView) {
        Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { i in
            // Update coordinates
            var currentCenter = mapView.centerCoordinate
            
            // Update coordinates
            currentCenter.latitude += 0.001
            currentCenter.longitude += 0.001
            
            // Animate camera to new coordinates
            let camera = MLNMapCamera(lookingAtCenter: currentCenter, altitude: mapView.camera.altitude, pitch: mapView.camera.pitch, heading: mapView.camera.heading)
//            mapView.setCamera(camera, withDuration: 1, animationTimingFunction: CAMediaTimingFunction(name: .easeInEaseOut))
//            mapView.setCamera(camera, withDuration: 1, animationTimingFunction: CAMediaTimingFunction(name: .easeInEaseOut))
            mapView.fly(to: camera)
        }
    }
}
```

**Comment by louwers** [comment]
I think we're losing a lot of accuracy on Android somewhere.

```
constexpr double kEpsilon = 1e-9;
// To avoid flickering issue due to "zoom = 13.9999999..".
double roundForAccuracy(double x) {
    double round_x = std::round(x);
    double diff = std::abs(round_x - x);
    if (diff < kEpsilon && diff > 0) {
        return round_x;
    } else {
        return x;
    }
}
```

is meant to account for this problem but the zoom goes from 15.0 to 14.993487604307171 during the animation

**Comment by louwers** [comment]
@geolives-contact `animateCamera` uses `flyTo` which is implemented using the [Smooth and efficient zooming and panning](https://www.win.tue.nl/~vanwijk/zoompan.pdf) algorithm. So it zooms out a little, with longer distances and a longer animation duration this is more pronounced. The zooming out causes (re)placement.

This looks related: https://github.com/maplibre/maplibre-native/issues/16

**Comment by louwers** [comment]
@geolives-contact Yes I think it's a bug, with such a small zoom change I don't expect such dramatic label placement changes. I also don't see it on iOS for some reason.

I see this problem in other styles as well, so I don't think it's related to your style. The solution is probably in the placement code but it quite complex so it may be hard to find the problem.

**Comment by jgillich** [comment]
I had a look at this and arrived at the same solution as @louwers earlier. flyTo performs a slight zoom out in the animation, which causes the layers to be swapped briefly. Skipping the zoom out when zoom level remains unchanged seems like a reasonable solution to me.

iOS appears to perform standard animations using `easeTo` instead of `flyTo`, which does not use the Van Wijk algorithm. We could also make this change here:

https://github.com/maplibre/maplibre-native/blob/647636bf6115e058991dd644638ba6748c9cf4c6/platform/android/MapLibreAndroid/src/main/java/org/maplibre/android/maps/Transform.java#L167

But that might have undesired side effects, not sure.

AI generated report:

<details>
 
> # Label Blinking at Integer Zoom Levels During Camera Animation
> 
> ## Issue
> 
> [#2477](https://github.com/maplibre/maplibre-native/issues/2477): Since MapLibre v11, labels blink when `animateCamera` is called at exact integer zoom levels (15.0, 16.0, 17.0, etc.). The issue does not occur at fractional zoom (e.g. 16.4 or 17.1) or with `moveCamera`.
> 
> ## Root Cause
> 
> The `flyTo` method implements the Van Wijk optimal path algorithm, which intentionally dips zoom below the starting value during a panning animation to create a sense of speed. At integer starting zooms (e.g. 15.0), this dip crosses the `std::floor()` boundary, causing the tile pipeline to select a different tile zoom level mid-flight.
> 
> ### Step-by-step trace
> 
> 1. **`flyTo` computes the Van Wijk path** at `src/mbgl/map/transform.cpp:318-319`:
>    ```cpp
>    auto w = [=](double s) {
>        return std::cosh(r0) / std::cosh(r0 + rho * s);
>    };
>    ```
>    `w(s)` grows above 1 during the flight, representing zoom-out.
> 
> 2. **Frame zoom is computed** at `src/mbgl/map/transform.cpp:367-368`:
>    ```cpp
>    double frameZoom = startZoom + state.scaleZoom(1 / w(s));
>    ```
>    With `w(s) > 1`, `1/w(s) < 1`, so `log2(1/w(s))` is negative—zooming below `startZoom`.
> 
> 3. **Zoom dips below the integer boundary**:
>    ```
>    startZoom=15.0  →  k=0.25: zoom=14.9914  →  k=0.50: zoom=14.9886  →  k=1.00: zoom=15.0000
>    ```
>    The dip is ~0.01 zoom levels (0.7% scale change), imperceptible to users.
> 
> 4. **`coveringZoomLevel` snaps to a different tile level** at `src/mbgl/util/tile_cover.cpp:149-155`:
>    ```cpp
>    int32_t coveringZoomLevel(double zoom, ...) noexcept {
>        return static_cast<int32_t>(std::floor(zoom)); // floor(14.9886) = 14
>    }
>    ```
>    Called from `src/mbgl/renderer/tile_pyramid.cpp:101`.
> 
> 5. **Tile level flips 15 → 14 → 15 during the animation**, causing:
>    - Different tile pyramid levels contain different label data, glyph sizes, and collision boxes
>    - The `CrossTileSymbolIndex` rematches symbols across zoom levels
>    - `holdForFade()` briefly retains wrong-level tiles
>    - The placement engine sees labels appear/disappear as tiles switch → visible blinking
> 
> ### Why it doesn't happen at fractional zoom
> 
> At zoom 15.4, the dip goes to ~15.39. `std::floor(15.39)` is still **15**. No tile level change, no blinking.
> 
> ### Why it doesn't happen with `moveCamera`
> 
> `moveCamera` calls `jumpTo` → `easeTo` with zero duration, setting the transform once rather than frame-by-frame. No flight path is computed.
> 
> ## Why iOS Is Not Affected
> 
> The bug only manifests through the `flyTo` code path with the Van Wijk curve. iOS avoids this entirely by calling `easeTo` for all standard camera animations.
> 
> ### Android → `flyTo`
> 
> At `platform/android/.../Transform.java:167`:
> ```java
> nativeMap.flyTo(cameraPosition.target, cameraPosition.zoom, ...);
> ```
> This calls `Transform::flyTo()` directly with `linearZoomInterpolation=false` (default), enabling the Van Wijk dip.
> 
> ### iOS → `easeTo`
> 
> At `platform/ios/src/MLNMapView.mm:4333`:
> ```objc
> self.mbglMap.easeTo(cameraOptions, animationOptions);
> ```
> `easeTo` uses linear zoom interpolation via `util::interpolate(startZoom, zoom, k)` — no dip.
> 
> All standard iOS animation APIs use `easeTo`:
> 
> | iOS API | Internal call |
> |---------|---------------|
> | `setCamera:animated:` | `easeTo` |
> | `setZoomLevel:animated:` | `easeTo` |
> | `setCenterCoordinate:zoomLevel:animated:` | `easeTo` |
> 
> iOS does have a `flyToCamera:` API that would exhibit the same bug, but it is a separate, explicitly-named method, not the default animation API.
> 
> ### Summary
> 
> | Platform | Method | C++ path | Van Wijk dip? | Blink? |
> |----------|--------|----------|---------------|--------|
> | Android | `animateCamera` | `flyTo` | Yes | Yes |
> | Android | `easeCamera` | `easeTo` | No | No |
> | iOS | `setCamera:animated:` | `easeTo` | No | No |
> | iOS | `flyToCamera:` | `flyTo` | Yes | Yes |
> 


</details>
